File: LinearList.java
class ListaLinear {
  public int[] items;
  public int max;
  public int n;

  public ListaLinear(int max) {
    this.items = new int[max];
    this.max = max;
    this.n = 0;
  }

  public void inserir(int elem, int pos) {
    if (this.n == this.max) throw new RuntimeException("Overflow");
    if (pos < 0 || pos > this.max) throw new RuntimeException("Position overflow");

    for (int i = this.n - 1; i >= pos; i--) {                
      this.items[i+1] = this.items[i];
    }
    this.items[pos] = elem;
    this.n++;
  }

  public void inserirInicio(int elem) {
    inserir(elem, 0);
  }

  public void inserirFim(int elem) {
    inserir(elem, this.n);
  }

  public void remover(int pos) {
    if (this.n == 0) throw new RuntimeException("Underflow");
    if (pos < 0 || pos > this.max) throw new RuntimeException("Position overflow");

    for (int i = pos; i < this.n; i++) {                
      this.items[i] = this.items[i+1];
    }
    this.items[this.n] = 0;
    this.n--;
  }

  public void removerInicio() {
    remover(0);
  }

  public void removerFim() {
    remover(this.n - 1);
  }

  public void mostrar() {
    for (int i = 0; i < this.n; i++) {
      System.out.println(this.items[i]);
    }
  }

  public boolean pesquisar(int elem) {
    for (int i = 0; i < this.n; i++) {
      if (this.items[i] == elem) return true;
    }
    return false;
  }
}

public class LinearList {
  public static void main(String args[]) {
    ListaLinear lista = new ListaLinear(10);
    lista.inserirFim(10);
    lista.inserirFim(20);
    lista.inserirFim(30);
    lista.inserirFim(40);
    lista.inserirFim(50);
    lista.mostrar();
    lista.removerInicio();
    lista.removerFim();
    lista.mostrar();
    System.out.println(lista.pesquisar(20));
  }
}

File: LinearList.c
#include <stdio.h>
#include <stdlib.h>

typedef struct ListaLinear
{
  int *items;
  int max;
  int n;
} ListaLinear;

ListaLinear newListaLinear(int max)
{
  ListaLinear lista;
  lista.items = (int *)malloc(max * sizeof(int));
  if (lista.items == NULL)
  {
    fprintf(stderr, "Erro: Falha na alocação de memória\n");
    exit(EXIT_FAILURE);
  }
  lista.max = max;
  lista.n = 0;
  return lista;
}

void delListaLinear(ListaLinear *lista)
{
  free(lista->items);
}

void inserirInicio(ListaLinear *lista, int elem)
{
  if (lista->n == lista->max)
  {
    fprintf(stderr, "Erro: Overflow\n");
    exit(EXIT_FAILURE);
  }

  for (int i = lista->n - 1; i >= 0; i--)
  {
    lista->items[i + 1] = lista->items[i];
  }
  lista->items[0] = elem;
  lista->n++;
}

void inserirFim(ListaLinear *lista, int elem)
{
  if (lista->n == lista->max)
  {
    fprintf(stderr, "Erro: Overflow\n");
    exit(EXIT_FAILURE);
  }

  lista->items[lista->n] = elem;
  lista->n++;
}

void inserir(ListaLinear *lista, int elem, int pos)
{
  if (lista->n == lista->max)
  {
    fprintf(stderr, "Erro: Overflow\n");
    exit(EXIT_FAILURE);
  }
  if (pos < 0 || pos > lista->max)
  {
    fprintf(stderr, "Erro: Posição inválida\n");
    exit(EXIT_FAILURE);
  }

  for (int i = lista->n - 1; i >= pos; i--)
  {
    lista->items[i + 1] = lista->items[i];
  }
  lista->items[pos] = elem;
  lista->n++;
}

int removerInicio(ListaLinear *lista)
{
  if (lista->n == 0)
  {
    fprintf(stderr, "Erro: Underflow\n");
    exit(EXIT_FAILURE);
  }

  int elem = lista->items[0];
  for (int i = 0; i < lista->n; i++)
  {
    lista->items[i] = lista->items[i + 1];
  }
  lista->n--;
  return elem;
}

int removerFim(ListaLinear *lista)
{
  if (lista->n == 0)
  {
    fprintf(stderr, "Erro: Underflow\n");
    exit(EXIT_FAILURE);
  }

  int elem = lista->items[lista->n - 1];
  lista->n--;
  return elem;
}

int remover(ListaLinear *lista, int pos)
{
  if (lista->n == 0)
  {
    fprintf(stderr, "Erro: Underflow\n");
    exit(EXIT_FAILURE);
  }
  if (pos < 0 || pos > lista->max)
  {
    fprintf(stderr, "Erro: Posição inválida\n");
    exit(EXIT_FAILURE);
  }

  int elem = lista->items[pos];
  for (int i = pos; i < lista->n; i++)
  {
    lista->items[i] = lista->items[i + 1];
  }
  lista->n--;
  return elem;
}

void mostrar(ListaLinear *lista)
{
  printf("Items: ");
  printf("[%d", lista->items[0]);
  for (int i = 1; i < lista->n; i++)
  {
    printf(", %d", lista->items[i]);
  }
  printf("]\n");
}

int pesquisar(ListaLinear *lista, int elem)
{
  for (int i = 0; i < lista->n; i++)
  {
    if (lista->items[i] == elem)
    {
      return 1;
    }
  }
  return 0;
}

int main()
{
  ListaLinear lista = newListaLinear(5);
  inserirInicio(&lista, 2);
  inserirInicio(&lista, 1);
  inserir(&lista, 3, 2);
  inserirFim(&lista, 4);

  mostrar(&lista);
  printf("Pesquisar 5: %d\n", pesquisar(&lista, 5));
  printf("Pesquisar 3: %d\n", pesquisar(&lista, 3));
  printf("Remover 3: %d\n", remover(&lista, 2));
  mostrar(&lista);
  printf("Remover Inicio: %d\n", removerInicio(&lista));
  mostrar(&lista);
  printf("Remover Fim: %d\n", removerFim(&lista));
  mostrar(&lista);
  delListaLinear(&lista);
  return 0;
}
